pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -17,20 +17,59 @@ limitations under the License.
 package persistentvolume
 
 import (
+	"fmt"
 	"time"
 
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/client/cache"
 	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
 	"k8s.io/kubernetes/pkg/cloudprovider"
 	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/conversion"
 	"k8s.io/kubernetes/pkg/runtime"
 	vol "k8s.io/kubernetes/pkg/volume"
 	"k8s.io/kubernetes/pkg/watch"
 
 	"github.com/golang/glog"
 )
 
+// Design:
+//
+// The fundamental key to this design is the bi-directional "pointer" between
+// PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs), which is
+// represented here as pvc.Spec.VolumeName and pv.Spec.ClaimRef. The bi-directionality
+// is complicated to manage in a transactionless system, but without it we
+// can't ensure sane behavior in the face of different forms of trouble.  For
+// example, a rogue HA controller instance could end up racing and making
+// multiple bindings that are indistinguishable, resulting in potential data
+// loss.
+//
+// This controller supports pre-bound (by the creator) objects in both
+// directions: a PVC that wants a specific PV or a PV that is reserved for a
+// specific PVC.
+//
+// The binding is two-step process. PV.Spec.ClaimRef is modified first and
+// PVC.Spec.VolumeName second. At any point of this transaction, the PV or PVC
+// can be modified by user or other controller or completelly deleted. Also, two
+// (or more) controllers may try to bind different volumes to different claims
+// at the same time. The controller must recover from any conflicts that may
+// arise from these conditions.
+
+// annWasEverBound annotation applies to PVCs.  It indicates that the lifecycle
```

Comment:
Comment that the existence of this annotation is the flag we need, the value is just expected to be non-empty.

I want to propose a simplification, and you can tell me to buzz off or to wait until this is in or maybe MAYBE add it as a commit at the end.

We have 2 annotations, both of which are essentially valueless existence flags.  I looked at pvhaxxz and I think I convinced myself these could be a single annotation with a value.

```
const annBoundBy = "pv.kubernetes.io/bound-by"

const boundByUser = "User"
const boundByController = "Controller"
```

As is (in pvhaxxz), when we detect that a claim was bound by user we set "was ever bound", but leave "bound by controller" off.  When we made the binding ourselves, we set "was ever bound" and "bound by controller" both.  Why not make "was ever bound" a bit more explicit and make the value meaningful.  We could define that any value other that "Controller" implies "User" and this controller logic will not change the binding.